captum/optim/_core/loss.py

```diff
@@ -138,18 +170,110 @@ def loss_fn(module: ModuleOutputMapping) -> torch.Tensor:
+
+        name = self.__name__
+        target = self.target
+    elif isinstance(other, Loss):
+        # This should never get called because __math_op__ will be called instead
+        pass
+    else:
+        raise TypeError(
+            "Can only apply math operations with int, float or Loss. Received type "
+            + str(type(other))
+        )
+    return CompositeLoss(loss_fn, name=name, target=target)
+
+
+def custom_composable_op(
+    loss,
+    loss_op_fn: Callable,
+    *args: Any,
+    **kwargs: Any,
+) -> "CompositeLoss":
+    """
+    Implement composability for operations that take a single tensor or list of tensors
+    and then return a single tensor. Custom user defined functions can be used in
+    addition to some built-in Python functions and PyTorch operations.
+
+    Args:
+
+        loss (Loss or list of Loss): A loss objective or list of loss objectives.
+        loss_op_fn (Callable): A supported PyTorch, Python, or custom function.
+            Default: torch.mean
+        args (Any, optional): Any additional arguments to pass to loss_op_fn.
+        kwargs (Any, optional): Any additional arguments to pass to loss_op_fn.
+        to_scalar_fn (Callable, optional): A function for converting loss function
+            outputs to scalar values, in order to prevent size mismatches. This is
+            variable only used if more than one loss is given.
+            Default: None
+
+        Returns:
+            composite_loss (CompositeLoss): A composable loss instance.
+    """
+
+    if isinstance(loss, (tuple, list)):
+        if "to_scalar_fn" not in kwargs:
+            to_scalar_fn = None
+        else:
+            to_scalar_fn = kwargs["to_scalar_fn"]
+            del kwargs["to_scalar_fn"]
+
+        def loss_fn(module: ModuleOutputMapping) -> torch.Tensor:
+            loss_tensors = [loss_obj(module) for loss_obj in loss]
+            if to_scalar_fn is not None:
+                loss_tensors = [to_scalar_fn(tensor) for tensor in loss_tensors]
+            return loss_op_fn(loss_tensors, *args, **kwargs)
+
+        name_list = ", ".join([loss_obj.__name__ for loss_obj in loss])
+        name = loss_op_fn.__name__ + "(" + name_list + ")"
+
+        # Collect targets from losses
+        target = [
+            target
+            for targets in [
+                [loss_obj.target]
+                if not hasattr(loss_obj.target, "__iter__")
+                else loss_obj.target
+                for loss_obj in loss
+            ]
+            for target in targets
+        ]
+    else:
+
+        def loss_fn(module: ModuleOutputMapping) -> torch.Tensor:
+            return loss_op_fn(loss(module), *args, **kwargs)
+
+        name = loss_op_fn.__name__ + "(" + loss.__name__ + ")"
+        target = loss.target
+    return CompositeLoss(loss_fn, name=name, target=target)
+
+
 class BaseLoss(Loss):
     def __init__(
         self,
         target: Union[nn.Module, List[nn.Module]] = [],

```

Comment:
can the `target` be `List[nn.Module]`? Many losses below directly access `targets_to_values[self.target]`. Did I miss anything?

Reply:
@aobo-y Losses like `ActivationInterpolation` have multiple targets (Faceted loss as well in an upcoming PR), but `BaseLoss` works off using a single `target` variable. 
 
The `BaseLoss` class is called in the `__init__` functions of loss classes like so:

```
# Single target
BaseLoss.__init__(self, target, batch_index)

# Multiple targets
BaseLoss.__init__(self, [target1, target2])
```

The loss class itself will indicate via `target: List[nn.Module]` type hint that multiple targets are supported / required, or it is handled things internally by passing the targets as a list to `BaseLoss` like in `ActivationInterpolation`.

The `ActivationInterpolation` loss class can be found here: https://github.com/ProGamerGov/captum/blob/optim-wip-composable-loss-improvements/captum/optim/_core/loss.py#L506

Reply:
Sure, but cases like `DeepDream` and some others directly inherits BaseLoss's init definition, where `target` can be a list while actually it should not https://github.com/ProGamerGov/captum/blob/optim-wip-composable-loss-improvements/captum/optim/_core/loss.py#L393-L407 

If these losses have different assumptions of what their targets should be, why do we abstract the `target` into the base class. The base class `BaseLoss` does not need `target` anyway. Each class can define their own `target` in `__init__`. Or we can have 2 other intermediate abstract classes `SingleTargetLoss` `MultiTargetsLoss`

But anyway, this is just for discussion. It has nth related to this PR. We can leave it to future updates if needed.

Reply:
Oh, yeah I see what you mean now. In the original code, I think that Ludwig had `SingleTargetObjective` & `MultiObjective` for handling these cases: https://github.com/ludwigschubert/captum/blob/f1fd0729dece59564a7c10b7b397617d8a09a247/captum/optim/optim/objectives.py#L108

It'd probably be best to leave this to a future PR if decide on the changes